consensus/clique/clique.go

```diff
@@ -538,10 +537,13 @@ func (c *Clique) Prepare(chain consensus.ChainReader, header *types.Header) erro
 		c.lock.RUnlock()
 	}
 	// Set the correct difficulty
-	header.Difficulty = diffNoTurn
-	if snap.inturn(header.Number.Uint64(), c.signer) {
-		header.Difficulty = diffInTurn
+	parent := chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
+	if parent == nil {
+		return consensus.ErrUnknownAncestor
 	}
+
+	header.Difficulty = c.CalcDifficulty(chain, 0, parent)
```

Comment:
Calling `CalcDifficulty` here is a bit expensive as it needs to regenerate the snapsot internally. Imho it would be better to leave this code as is and calculate the difficulty directly.

Alternatively, you can add a global `clique.CalcDifficulty` method the same was as in `ethash.CalcDifficulty`, that would contain just the essence and operate on a snapshot:

```go
func CalcDifficulty(snap *Snapshot, signer common.Address) *big.Int {
	if snap.inturn(snap.Number+1, c.signer) {
		return new(big.Int).Set(diffInTurn)
	}
	return new(big.Int).Set(diffNoTurn)
}
```

And you can call this function firectly here, and also wrap it with engine.CalcDifficulty.